examples/consistency_models/train_consistency_distillation.py

```diff
@@ -0,0 +1,701 @@
+    max_train_steps = args.num_epochs * num_update_steps_per_epoch
+
+    logger.info("***** Running training *****")
+    logger.info(f"  Num examples = {len(dataset)}")
+    logger.info(f"  Num Epochs = {args.num_epochs}")
+    logger.info(f"  Instantaneous batch size per device = {args.train_batch_size}")
+    logger.info(f"  Total train batch size (w. parallel, distributed & accumulation) = {total_batch_size}")
+    logger.info(f"  Gradient Accumulation steps = {args.gradient_accumulation_steps}")
+    logger.info(f"  Total optimization steps = {max_train_steps}")
+
+    global_step = 0
+    first_epoch = 0
+
+    # Potentially load in the weights and states from a previous save
+    if args.resume_from_checkpoint:
+        if args.resume_from_checkpoint != "latest":
+            path = os.path.basename(args.resume_from_checkpoint)
+        else:
+            # Get the most recent checkpoint
+            dirs = os.listdir(args.output_dir)
+            dirs = [d for d in dirs if d.startswith("checkpoint")]
+            dirs = sorted(dirs, key=lambda x: int(x.split("-")[1]))
+            path = dirs[-1] if len(dirs) > 0 else None
+
+        if path is None:
+            accelerator.print(
+                f"Checkpoint '{args.resume_from_checkpoint}' does not exist. Starting a new training run."
+            )
+            args.resume_from_checkpoint = None
+        else:
+            accelerator.print(f"Resuming from checkpoint {path}")
+            accelerator.load_state(os.path.join(args.output_dir, path))
+            global_step = int(path.split("-")[1])
+
+            initial_global_step = global_step * args.gradient_accumulation_steps
+            first_epoch = global_step // num_update_steps_per_epoch
+    else:
+        initial_global_step = 0
+
+    progress_bar = tqdm(
+        range(0, max_train_steps),
+        initial=initial_global_step,
+        desc="Steps",
+        # Only show the progress bar once on each machine.
+        disable=not accelerator.is_local_main_process,
+    )
+
+
+
+    # Train!
+    for epoch in range(first_epoch, args.num_epochs):
+        model.train()
+        for step, batch in enumerate(train_dataloader):
+            clean_images = batch["input"]
+            labels = batch["labels"]
+            # Sample noise that we'll add to the images
+            noise = torch.randn(clean_images.shape).to(clean_images.device)
+            # Sample a random timestep for each image, TODO - allow different timesteps in a batch
+            index = torch.randint(
+                0, noise_scheduler.config.num_train_timesteps-1,  (1,), device=clean_images.device
+            ).long()
+            timestep = timesteps[index]
+            timestep_prev = timestep + 1
+            # TODO, we should have an add noise in the scheduler maybe?
+            noised_image = clean_images + noise*append_dims(timestep, clean_images.ndim)
```

Comment:
I will add a `add_noise` method in the scheduler that implements this step (e.g. sampling a noised version of the image $`x_{t_{n+ 1}}`$ from $`\mathcal{N}(x; t_{n + 1}^2\textbf{I})`$, following Line 4 of Algorithm 2).

As an aside, I think `timestep` here should be `noise_scheduler.sigmas[index]` rather than `noise_scheduler.timesteps[index]`. My understanding is that whenever the paper talks about timesteps (e.g. $`\{\tau_n\}`$ in Algorithm 1 or $`\{t_n\}`$ in Algorithm 2) it's typically referring to timesteps already mapped to Karras sigmas (in decreasing and increasing order, respectively). It's confusing because the Karras sigma schedule can also be thought of timesteps which break up the time interval $`[\epsilon, T] = [\sigma_{min}, \sigma_{max}]`$ into unequally sized subintervals that become smaller as we go backward in time (see Section 4 of the [paper](https://arxiv.org/pdf/2303.01469.pdf)).

Reply:
Added.